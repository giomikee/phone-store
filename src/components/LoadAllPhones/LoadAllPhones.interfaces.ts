export interface Props {
    onLoadAll: () => void;
}