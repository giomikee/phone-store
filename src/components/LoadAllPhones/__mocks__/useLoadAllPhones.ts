import { vi } from 'vitest';

export const useLoadAllPhones = vi.fn().mockReturnValue({
    isLoading: false,
    fetchAllPhones: vi.fn(),
});