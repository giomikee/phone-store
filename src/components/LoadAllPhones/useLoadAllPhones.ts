import { useDispatch } from 'react-redux';
import { useNavigate } from 'react-router-dom';
import { INITIAL_PHONES_COUNT, PAGES } from '../../constants';
import { useState } from 'react';
import { getPhones } from '../../services';
import { addPhones } from '../../state/phone/phone.slice';
import { useStoreSelectors } from '../../state/useStoreSelectors';

export function useLoadAllPhones(onLoadAll: () => void) {
    const navigate = useNavigate();
    const dispatch = useDispatch();

    const { phones } = useStoreSelectors();
    const hasAllPhones = phones.length > INITIAL_PHONES_COUNT;

    const [isLoading, setIsLoading] = useState(false);

    const fetchAllPhones = async () => {
        if (isLoading || hasAllPhones) {
            if (hasAllPhones) {
                onLoadAll();
            }

            return;
        }

        setIsLoading(true);

        try {
            const fetchedPhones = await getPhones({ offset: INITIAL_PHONES_COUNT });
            dispatch(addPhones(fetchedPhones));
        } catch (error) {
            console.error(error);
            navigate(PAGES.error);
        } finally {
            setIsLoading(false);
            onLoadAll();
        }
    };

    return {
        isLoading,
        setIsLoading,
        fetchAllPhones,
    };
}