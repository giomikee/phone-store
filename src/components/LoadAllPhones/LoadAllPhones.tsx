import ExpandCircleDownIcon from '@mui/icons-material/ExpandCircleDown';
import { Box, Button } from '@mui/material';
import type { Props } from './LoadAllPhones.interfaces';
import { useLoadAllPhones } from './useLoadAllPhones';

const LoadAllPhones = ({onLoadAll}: Props) => {
    const {isLoading, fetchAllPhones} = useLoadAllPhones(onLoadAll);

    return (
        <Box sx={{ flexGrow: 1, textAlign: 'center'}}>
            <Button 
                data-testid='load-all-phones-button'
                loading={isLoading}
                onClick={fetchAllPhones}
                startIcon={<ExpandCircleDownIcon />}
                variant='contained'
            >
                Ver todo
            </Button>
        </Box>
    )
}

export default LoadAllPhones