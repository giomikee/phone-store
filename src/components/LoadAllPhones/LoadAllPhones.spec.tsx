import { describe, expect, it, vi, type Mock } from 'vitest';
import { render, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom/vitest';
import LoadAllPhones from './LoadAllPhones';
import { useLoadAllPhones } from './useLoadAllPhones';


vi.mock('./useLoadAllPhones');

describe('Given LoadAllPhones component', () => { 
    const defaultProps = {
        onLoadAll: vi.fn(),
    };

    describe('When the component is rendered', () => {
        it('Then it should match the snapshot', () => {
            const { container } = render(<LoadAllPhones {...defaultProps} />);
            expect(container).toMatchSnapshot();
        });
    });

    describe('When load all phones button is clicked', () => {
        it('Then fetchAllPhones should be called', () => {
            const fetchAllPhonesMock = vi.fn();

            (useLoadAllPhones as Mock).mockReturnValueOnce({
                isLoading: false,
                fetchAllPhones: fetchAllPhonesMock,
            })

            render(<LoadAllPhones {...defaultProps} />);

            fireEvent.click(screen.getByTestId('load-all-phones-button'));
            expect(fetchAllPhonesMock).toHaveBeenCalled();
        });
    });
})