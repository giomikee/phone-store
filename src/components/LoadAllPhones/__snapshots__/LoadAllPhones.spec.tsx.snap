// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Given LoadAllPhones component > When the component is rendered > Then it should match the snapshot 1`] = `
<div>
  <div
    class="MuiBox-root css-17ls7xt"
  >
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-74d805-MuiButtonBase-root-MuiButton-root"
      data-testid="load-all-phones-button"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-icon MuiButton-startIcon MuiButton-iconSizeMedium css-1sh91j5-MuiButton-startIcon"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
          data-testid="ExpandCircleDownIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2m0 13.5L7.5 11l1.42-1.41L12 12.67l3.08-3.08L16.5 11z"
          />
        </svg>
      </span>
      <span
        class="MuiButton-loadingWrapper"
        style="display: contents;"
      />
      Ver todo
    </button>
  </div>
</div>
`;
